tests/test_operator.py

```diff
@@ -203,6 +203,92 @@ def test_arguments(self):
         op4 = switchconfig(language='openmp')(Operator)(Eq(u, u + 1), language='C')
         assert '#pragma omp for' not in str(op4)
 
+    def test_nested_lowering_and_indexing(self):
+        """
+        Tests that deeply nested (depth > 2) functions are lowered and indexified.
+        """
+        grid = Grid(shape=(4, 4))
+        x, y = grid.dimensions
+
+        u1 = Function(name="u1", grid=grid, dtype=np.int32)
+        u2 = Function(name="u2", grid=grid, dtype=np.int32)
+        u3 = Function(name="u3", grid=grid, dtype=np.int32)
+        u4 = Function(name="u4", grid=grid, dtype=np.int32)
+        u5 = Function(name="u5", grid=grid, dtype=np.int32)
+
+        Eq1 = Eq(u1, u2[x, u3[x, u4[x, u5[x, u1]]]])
+        Eq2 = Eq(u1, u2[x, u3[x, u4[x, u5[x, u1[x, y]]]]])
+
+        op1 = Operator([Eq1])
+        op2 = Operator([Eq2])
+
+        op1.apply()
+        op2.apply()
+        assert ('u1[x + 1][y + 1] = u2[x + 1][u3[x + 1][u4[x + 1][u5[x + 1]'
+                '[u1[x + 1][y + 1] + 1] + 1] + 1] + 1]') in str(op2.ccode)
+
+        assert str(op2.ccode) == str(op1.ccode)
+
+    def test_nested_lowering_and_indexing_rhs_lhs(self):
```

Comment:
what's the difference between this one and the last test? conceptually, I mean

Reply:
and what's the difference w.r.t. the first test?

Reply:
Double indirection both at x and y + no grid, so as to have explicitly defined dimensions. Otherwise there exist cyclic dependencies.